src/api_test.ts

```diff
@@ -1,142 +1,296 @@
-        mf.mock("POST@/api/chat.postMessage", () => {
-          return new Response('{"ok":true}');
-        });
+        await t.step(
+          "should throw if response returns an HTTP status code >= 400",
+          async () => {
+            MockFetch.mock("POST@/api/chat.postMessage", () => {
+              return new Response("big explosions", { status: 500 });
+            });
+
+            await assertRejects(
+              async () => {
+                return await client.apiCall("chat.postMessage", {});
+              },
+              Error,
+              "500: big explosions",
+            );
+
+            MockFetch.reset();
+          },
+        );
+
+        await t.step(
+          "should throw an HttpError if HTTP response status code >= 400",
+          async () => {
+            MockFetch.mock("POST@/api/chat.postMessage", () => {
+              return new Response("ratelimited", {
+                status: 429,
+                headers: { "Retry-After": "120" },
+              });
+            });
+
+            await assertRejects(
+              async () => {
+                return await client.apiCall("chat.postMessage", {});
+              },
+              (error: Error) => {
+                assertInstanceOf(error, HttpError);
+                if (isHttpError(error)) {
+                  assertEquals(error.headers?.get("Retry-After"), "120");
+                  assertEquals(error.status, 429);
+                  assertEquals(error.message, "429: ratelimited");
+                }
+              },
+            );
+
+            MockFetch.reset();
+          },
+        );
+
+        await t.step(
+          "should return successful response JSON",
+          async () => {
+            MockFetch.mock("POST@/api/chat.postMessage", () => {
+              return new Response('{"ok":true}');
+            });
+
+            const res = await client.apiCall("chat.postMessage", {});
+            assertEquals(res.ok, true);
 
-        const res = await client.apiCall("chat.postMessage", {});
-        assertEquals(res.ok, true);
+            MockFetch.reset();
+          },
+        );
+
+        await t.step(
+          "should return usable Response with payload {'ok': false}",
+          async () => {
+            MockFetch.mock("POST@/api/chat.postMessage", () => {
+              return new Response('{"ok":false}', {
+                headers: { "Retry-After": "120" },
+              });
+            });
+
+            const res = await client.apiCall("chat.postMessage", {});
+            assertEquals(res.ok, false);
+            const fullRes = res.toFetchResponse();
+            assertInstanceOf(fullRes, Response);
+            assertEquals(fullRes.headers?.get("Retry-After"), "120");
+
+            MockFetch.reset();
+          },
+        );
 
-        mf.reset();
+        await t.step(
+          "should return usable Response with payload {'ok': false}",
```

Comment:
Because of the way the Slack API works the `HTTP status` of a request can be `200` but the payload indicates that the "Slack task" failed by providing `"ok":false` in the payload

Sometime a user may want to access the `headers` of a request that has `HTTP status` = `200` and payload with  `"ok":false` this test ensure that they are able since no `HttpError` is raised

lmk if this makes sense